floodmodeller_api/toolbox/model_build/structure_log/structure_log.py

```diff
@@ -1,14 +1,48 @@
+import copy
 import csv
+import json
 
 from floodmodeller_api import DAT
 
 
+def extract_attrs(object_input, keys):
+    """
+    extract attributes and their values from a given object and return as a dict
+
+    No handling for where attributes dont exist but can ignore for when this is an issue, shouldnt be an issue here
+    """
+    return {key: getattr(object_input, key) for key in keys}
+
+
+def serialise_keys(old_dict):
+    """
+    dictionary keys must be unique; but in FM the node label isnt always unique
+    however the (label,type) pair will always be unique; so we have to use that as our key.
+    json doesnt like tuples as keys so we have to convert them first if we need to serialise our dictionary output.
+
+    Replaces tuple keys with string versions of themselves, wrapped in ()
+
+    """
+    new_dict = {}
+    for key, value in old_dict.items():
+        new_dict[f'({",".join(key)})'] = value
+
+    return new_dict
+
+
 class StructureLogBuilder:
-    def __init__(self, input_path, output_path) -> None:
+
+    def __init__(self, input_path="", output_path="") -> None:
         self.dat_file_path = input_path
         self.csv_output_path = output_path
+        self.conduit_chains = (
+            {}
+        )  # pylint flags these for type hinting, but not sure how to specify, to discuss
```

Comment:
Is this just an empty dictionary? (i.e. not sure if you were trying to define a tuple of dictionaries, as that would require an extra comma). If I remember correctly, mypy likes you to typehint a dictionary when defining it so we know what type of dictionary it is. So you would write something like e.g. `self.conduit_chains: dict[str, list[str]] = {}` if it was destined to be a dictionary of lists of strings

Reply:
I did add this exact typehinting in a later commit, the parenthesis and new line was something that black insisted on adding.